src/compute-types/src/dataflows.rs

```diff
@@ -67,19 +68,26 @@ pub struct DataflowDescription<P, S: 'static = (), T = mz_repr::Timestamp> {
 impl<T> DataflowDescription<Plan<T>, (), mz_repr::Timestamp> {
     /// Tests if the dataflow refers to a single timestamp, namely
     /// that `as_of` has a single coordinate and that the `until`
-    /// value corresponds to the `as_of` value plus one.
+    /// value corresponds to the `as_of` value plus one, or `as_of`
+    /// is the maximum timestamp and is thus single.
     pub fn is_single_time(&self) -> bool {
         // TODO: this would be much easier to check if `until` was a strict lower bound,
         // and we would be testing that `until == as_of`.
         let Some(as_of) = self.as_of.as_ref() else {
             return false;
         };
-        !as_of.is_empty()
-            && as_of
-                .as_option()
-                .and_then(|as_of| as_of.checked_add(1))
-                .as_ref()
-                == self.until.as_option()
+        let Some(as_of) = as_of.as_option() else {
+            return false;
+        };
+        // Ensure that as_of = MAX implies until.is_empty().
+        soft_assert_or_log!(
+            as_of != &mz_repr::Timestamp::MAX || self.until.is_empty(),
+            "expected empty until due to MAX as_of, got {:?}",
+            self.until
+        );
```

Comment:
Given the comment below, this `soft_assert_or_log!` is standing in more as a form of input validation and not as part of the core logic to determine if the dataflow is single-time, correct? In this case, it would be good to extend the condition a bit, e.g., by also validating that the `until` does not precede the `as_of`.

Reply:
I added a new condition.

Reply:
Perfect, thanks!